core/arch/arm/kernel/thread.c

```diff
@@ -223,6 +239,21 @@ void thread_check_canaries(void)
 #endif/*CFG_WITH_STACK_CANARIES*/
 }
 
+#ifdef CFG_CORE_CHECK_STACKS
+void __cyg_profile_func_enter(void *this_fn, void *call_site);
+void __noprof __cyg_profile_func_enter(void *this_fn __unused,
+				       void *call_site __unused)
+{
+}
+
+void __cyg_profile_func_exit(void *this_fn, void *call_site);
+void __noprof __cyg_profile_func_exit(void *this_fn __unused,
+				      void *call_site __unused)
+{
+	thread_check_canaries();
```

Comment:
So this happens at the end of each function call? I'm wondering if there's some trick we could use to only check the current stack.

Reply:
Yes, this is clearly sub-optimal. We could perhaps do something like:
```C
thr_id = thread_get_id_may_fail();
if (thr_id >= 0) {
        check_sp_is_in_thread_stack(sp, thr_id);
} else if (using_tmp_stack()) {     /* <=== How do you implement this line? */
        check_sp_is_in_tmp_stack(sp, core_id);
} else {
        check_sp_is_in_abt_stack(sp, core_id);
}
```
We would not even need to check the canaries, checking the stack pointer would be enough. So no dependency on `CFG_WITH_STACK_CANARIES`.

But how to decide between tmp and abt stack?

Reply:
The `flags` field in `struct thread_core_local` should help with that.